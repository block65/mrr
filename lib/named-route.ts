import type { Params, ExtractRouteParams } from './types.js';
import { nullOrigin, urlObjectAssign, urlRhs } from './util.js';

type QueryParams = Record<string, string>;

export interface NamedRoute<P extends string, Q extends QueryParams> {
  path: P;
  build: (x?: {
    params?: ExtractRouteParams<P>;
    query?: Q;
    hash?: string;
    origin?: string;
  }) => string;
}

export function interpolate<P extends string>(
  path: P,
  params: ExtractRouteParams<P> | Params = {},
): string {
  return path.replace(/\/:(\w+)[?+*]?/g, (_match, token: keyof typeof params) =>
    params[token] ? `/${params[token]}` : '',
  );
}

export function namedRoute<T extends string, Q extends QueryParams = never>(
  pathname: T,
): NamedRoute<T, Q> {
  return {
    path: pathname,
    build({ params, query, origin, ...rest } = {}) {
      const interpolatedPathname = params
        ? interpolate(pathname, params)
        : pathname;

      const searchParams = new URLSearchParams(query);
      searchParams.sort();

      const newUrl = urlObjectAssign(new URL(origin || nullOrigin), {
        pathname: interpolatedPathname,
        search: searchParams.toString(),
        ...rest,
      });

      return newUrl.origin === nullOrigin.origin
        ? urlRhs(newUrl)
        : newUrl.toString();
    },
  };
}
