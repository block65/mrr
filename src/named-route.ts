export { namedRoute } from '../lib/named-route.js';

export type {
  NamedRoute,
  NamedRouteWithParams,
  NamedRouteWithoutParams,
} from '../lib/named-route.js';
