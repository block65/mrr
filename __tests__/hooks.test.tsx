import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { expect, test } from 'vitest';
import type { ExtractRouteParams } from '../lib/types.js';
import { Route, Router, Routes, useRouteParams } from '../src/index.js';
import { namedRoute } from '../src/named-route.js';

const login = namedRoute('/foo/:foo/bar/:bar?');

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const EffCee = (_: ExtractRouteParams<typeof login.path>) => {
  const match = useRouteParams();
  return <pre>{JSON.stringify(match, null, 2)}</pre>;
};

test('Router preset pathname', async () => {
  const { asFragment } = render(
    <Router
      pathname={login.build({
        params: {
          foo: 'oh-look-a-foo',
          bar: 'omg-a-bar',
        },
      })}
    >
      <Routes>
        <Route path={login.path} component={EffCee} />
        <Route>if you see this, the test failed</Route>
      </Routes>
    </Router>,
  );

  expect(asFragment()).toMatchSnapshot();
});
