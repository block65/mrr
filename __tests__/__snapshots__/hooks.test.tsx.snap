// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Router preset pathname 1`] = `
<DocumentFragment>
  <pre>
    {
  "foo": "oh-look-a-foo",
  "bar": "omg-a-bar"
}
  </pre>
</DocumentFragment>
`;
